vpr/src/noc/noc_traffic_flows.h

```diff
@@ -0,0 +1,307 @@
+
+    /**
+     * @brief Given a router block in the clustered netlist, store it
+     * internally.
+     * 
+     * @param router_block_id A unique identifier that rerpesents a router block in the clustered netlist. This id will be stored.
+     */
+    void add_router_to_block_set(ClusterBlockId router_block_id);
+
+  public:
+    NocTrafficFlows();
+
+    //getters
+
+    /**
+     * @brief Given a unique id of a traffic flow (t_noc_traffic_flow)
+     * retrieve it from the vector of all traffic flows in the design. The
+     * retrieved traffic flow cannot be modified but can be used to
+     * retireve information such as the routers involved.
+     * 
+     * @param traffic_flow_id The unique identifier (NocTrafficFlowId)
+     * of the traffic flow to retrieve.
+     * @return const t_noc_traffic_flow& The traffic flow represented by
+     * the provided identifier.
+     */
+    const t_noc_traffic_flow& get_single_noc_traffic_flow(NocTrafficFlowId traffic_flow_id) const;
+
+    /**
+     * @brief Get a vector of all traffic flows that have a given router
+     * block in the clustered netlist as the source (starting point) in the 
+     * flow.
+     * 
+     * @param source_router_id A unique identifier that represents the
+     * a router block in the clustered netlist. This router block will
+     * be the source router in the retrieved traffic flow.
+     * @return const std::vector<NocTrafficFlowId>* A vector of traffic 
+     * flows that have the input router block parameter as the source in the
+     * flow.
+     */
+    const std::vector<NocTrafficFlowId>* get_traffic_flows_associated_to_source_router(ClusterBlockId source_router_id) const;
+
+    /**
+     * @brief Get a vector of all traffic flows that have a given router
+     * block in the clustered netlist as the sink (end point) in the 
+     * flow.
+     * 
+     * @param sink_router_id A unique identifier that represents the
+     * a router block in the clustered netlist. This router block will
+     * be the sink router in the retrieved traffic flow.
+     * @return const std::vector<NocTrafficFlowId>* A vector of traffic 
+     * flows that have the input router block parameter as the sink in the
+     * flow.
+     */
+    const std::vector<NocTrafficFlowId>* get_traffic_flows_associated_to_sink_router(ClusterBlockId sink_router_id) const;
+
+    /**
+     * @brief Given a traffic flow, determine whether it has been processed
+     * or not. By processed, it checks to see whether the traffic flow has
+     * been routed. This function should be used before routing a traffic
+     * flow.
+     * 
+     * @param traffic_flow_id A unique identifier that represents a traffic
+     * flow
+     * @return true Represents the case where the traffic flow has been
+     * processed
+     * @return false Represents the case where the traffic flow has not
+     * been processed 
+     */
+    bool get_traffic_flow_processed_status(NocTrafficFlowId traffic_flow_id);
```

Comment:
Evaluate if this can be done with local vars (and an unordered_set)

Reply:
Decided to use a local unordered_set that will be created during each iteration of placement. And whenever a traffic flow is routed, its id will be added to this set. Then this set will be used to check whether a traffic flow was already routed or not.